Managers/UInputMan.h

```diff
@@ -1,1684 +1,815 @@
+
+		/// <summary>
+		/// Sets whether a mouse button is pressed by a player during network multiplayer.
+		/// </summary>
+		/// <param name="player">Which player to set for.</param>
+		/// <param name="whichButton">Which mouse button to set for.</param>
+		/// <param name="state">The new state of the mouse button. True or false.</param>
+		void SetNetworkMouseButtonPressedState(short player, int whichButton, bool state);
+
+		/// <summary>
+		/// Sets whether a mouse button is released by a player during network multiplayer.
+		/// </summary>
+		/// <param name="player">Which player to set for.</param>
+		/// <param name="whichButton">Which mouse button to set for.</param>
+		/// <param name="state">The new state of the mouse button. True or false.</param>
+		void SetNetworkMouseButtonReleasedState(short player, int whichButton, bool state);
+
+		/// <summary>
+		/// Sets the state of the mouse wheel for a player during network multiplayer.
+		/// </summary>
+		/// <param name="player">The player to set for.</param>
+		/// <param name="state">The new state of the mouse wheel.</param>
+		void SetNetworkMouseWheelState(short player, int state);
+
+		/// <summary>
+		/// Gets whether the specified input element is pressed during network multiplayer.
+		/// </summary>
+		/// <param name="element">The input element to check for.</param>
+		/// <returns>Whether the specified input element is pressed or not.</returns>
+		bool AccumulatedElementPressed(int element) { return (element < 0 || element >= INPUT_COUNT) ? false : m_NetworkAccumulatedElementPressed[element]; }
+
+		/// <summary>
+		/// Gets whether the specified input element is released during network multiplayer.
+		/// </summary>
+		/// <param name="element">The input element to check for.</param>
+		/// <returns>Whether the specified input element is released or not.</returns>
+		bool AccumulatedElementReleased(int element) const { return (element < 0 || element >= INPUT_COUNT) ? false : m_NetworkAccumulatedElementReleased[element]; }
+
+		/// <summary>
+		/// Clears all the accumulated input element states.
+		/// </summary>
+		void ClearAccumulatedStates();
+#pragma endregion
+
+#pragma region Class Info
+		/// <summary>
+		/// Gets the class name of this object.
+		/// </summary>
+		/// <returns>A string with the friendly-formatted type name of this object.</returns>
+		virtual const std::string & GetClassName() const { return c_ClassName; }
+#pragma endregion
+
+	protected:
+
+		static const std::string c_ClassName; //!< A string with the friendly-formatted type name of this object.
+
+		static GUIInput* s_InputClass; //!< Current input class if available.
+
+		static char *s_PrevKeyStates; //!< Key states as they were the previous update.
+		static char *s_ChangedKeyStates; //!< Key states that have changed.
+
+		//TODO: Figure out why we need this for mouse but not for other input devices.
+		static bool s_MouseButtonStates[MAX_MOUSE_BUTTONS]; //!< Current? mouse states.
```

Comment:
I couldn't really figure out why there's 3 arrays to deal with the mouse while other devices do it with 2.
I have a feeling one of these is garbage, but it also might be to compensate some other bullshit.

Reply:
Ohh, I see what you're saying now, not why we have an array of 3 but why we have s_MouseButtonStates, PreviousStates and ChangedStates instead of just PreviousStates and ChangedStates?

I'm pretty sure we don't need it, it's just cleaner. If we wanna check if mouse button 1 is held this frame, we'd do `mouse1Held = mouse_b & 1;` (note that & is bitwise and). If we wanna check that joy button 1 is held, we'd do `joy1Held = joy[whichJoy].button[whichButton].b`. The mouse one just puts it in a variable cause it's nicer to work with (I think).

To be honest, it might be nice to handle all of them like we do mouse button; it'd increase readability and mean we only need to query allegro input once per frame instead of multiple times.